Fix uses of require in goroutines

zmb3 opened the pull request:
The require checks from testify call (*testing.T).Fail, which should only be called from the test goroutine.

zmb3 committed:
Fix uses of require in goroutines

The require checks from testify call (*testing.T).Fail, which should
only be called from the test goroutine.

codingllama commented on api/client/client_test.go:
```diff
@@ -104,7 +104,7 @@ func startMockServerWithListener(t *testing.T, l net.Listener) *mockServer {
 
 	go func() {
-		require.NoError(t, srv.grpc.Serve(l))
+		srv.grpc.Serve(l)
```
FYI, you can use assert.NoError - the problem is calling `t.Fatal()` (directly or indirectly) in goroutines, calling `t.Error` (and variants) is alright.

In this case I think keeping the assert is good, so we can narrow down mysterious failures to gRPC not starting.

```suggestion
		assert.NoError(t, srv.grpc.Serve(l))
```

zmb3 replied:
Yeah, though it's still racy, if the main test goroutine finishes before this one, the assert will be missed and an error from `Serve` won't fail the test. The proper way to catch this if you really care about `Serve` not returning an error is to dump the result to a channel and check it when the test finishes, but that's hard to do with an asynchronous helper like this.

In this case, I figured if the server fails to start, the test will fail on a future assertion.

codingllama replied:
> In this case, I figured if the server fails to start, the test will fail on a future assertion

I think this is exactly when it makes a difference, especially when one goes about debugging a test they haven't seen before. Up to you.

codingllama reviewed (commented):
Let me know if you'd rather keep the channels or change other calls to use assert.

codingllama commented on lib/backend/firestore/firestorebk_test.go:
```diff
@@ -270,7 +270,10 @@ func TestDeleteDocuments(t *testing.T) {
+
+			errCh := make(chan error)
+			go func() { errCh <- srv.Serve(lis) }()
+			t.Cleanup(func() { require.NoError(t, <-errCh) })
```
Should we do a single cleanup that does Stop and then receives from errCh? It's a bit more resistant to eventual changes.

zmb3 replied:
We should be able do that so long as we buffer the channel. Otherwise the stacked behavior of cleanups would create a deadlock. Let me give it a try.

zmb3 committed:
Address review comments

- Restore grpc.Serve() error check
- Combine the stop and channel read into a single cleanup

codingllama commented on lib/backend/firestore/firestorebk_test.go:
```diff
@@ -270,8 +270,13 @@ func TestDeleteDocuments(t *testing.T) {
-			go func() { require.NoError(t, srv.Serve(lis)) }()
-			t.Cleanup(srv.Stop)
+
+			errCh := make(chan error)
```
```suggestion
			errCh := make(chan error, 1)
```

Should this one be buffered too?

zmb3 replied:
I think that would be safest! Good catch.

zmb3 committed:
Merge branch 'master' into zmb3/fix-async-requires

zmb3 committed:
Merge branch 'master' into zmb3/fix-async-requires
